### ISSUE TITLE
Update the current Scheduler api?

### ISSUE THREAD

**Issue by He-Pin** [typelevel/cats-effect#3356]
Hi I am still new on cats-effect ,but I think the current api of Scheduler can be updated with :
```scala
trait Scheduler {
  type TimerId= Long

  def scheduleOnce(delay: FiniteDuration, task: Runnable): TimerId

   def scheduleWithFixedDelay(initialDelay:FiniteDuration, delay: FiniteDuration, task: Runnable):TimerId

   def scheduleAtFixedRate(initialDelay:FiniteDuration, delay: FiniteDuration, task: Runnable):TimerId

   def cancel(timerId:TimerId):Boolean

   def clock(): Clock
   ...
  ...
}
```

And seems like the only way to cancel the current scheduled task is by call the returning `Runnable` with the current Scheduler api?

As I'm starting using vertx these days, and I think the `cats.effect.unsafe.WorkerThread` can do the scheduling too, just as Netty's Eventloop does, which employed an PriorityQueue to do that.

**Comment by armanbilge** [comment]
> And seems like the only way to cancel the current scheduled task is by call the returning `Runnable` with the current Scheduler api?

That's correct. Unfortunately, we cannot change the API at this point due to compatibility constraints.

--- 

> As I'm starting using vertx these days, and I think the `cats.effect.unsafe.WorkerThread` can do the scheduling too, just as Netty's Eventloop does, which employed an PriorityQueue to do that.

It can, and it will very soon! See this PR:
- https://github.com/typelevel/cats-effect/pull/3219

**Comment by He-Pin** [comment]
And then, a user can using a HashWheelTimer which turn out the timeout as an event and send to the workerThread's taskQueue.

Update:
Thanks, I will take sometime to learn from this new PR.

**Comment by lqhuang** [comment]
> That's correct. Unfortunately, we cannot change the API at this point due to compatibility constraints.

@armanbilge May I have a question about when or what conditions are satisfied, then we could make improvements and break compatibility constraints.

Thanks!

**Comment by armanbilge** [comment]
> May I have a question about when or what conditions are satisfied, then we could make improvements and break compatibility constraints.

When we are working on Cats Effect 4, which would be a major, breaking change. But there are no concrete plans for this at the moment. It would probably happen around the same time as Cats 3, which is also rather hypothetical.

- https://github.com/typelevel/cats/issues/3757
